lib/compat/wordpress-6.6/resolve-patterns.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+if ( ! function_exists( 'gutenberg_replace_pattern_blocks' ) ) {
+	/**
+	 * Replaces pattern blocks with their content.
+	 *
+	 * @param array $blocks Array of blocks.
+	 * @param array $inner_content Optional array of inner content.
+	 * @return array Array of blocks with patterns replaced.
+	 */
+	function gutenberg_replace_pattern_blocks( $blocks, &$inner_content = null ) {
+		$i = 0;
+		// Also process new blocks that are inserted.
+		while ( $i < count( $blocks ) ) {
+			if ( 'core/pattern' === $blocks[ $i ]['blockName'] ) {
+				$registry         = WP_Block_Patterns_Registry::get_instance();
+				$pattern          = $registry->get_registered( $blocks[ $i ]['attrs']['slug'] );
+				$pattern_content  = $pattern['content'];
+				$blocks_to_insert = parse_blocks( $pattern_content );
+				array_splice( $blocks, $i, 1, $blocks_to_insert );
+
+				// If we have inner content, we need to insert nulls in the
+				// inner content array, otherwise serialize_blocks will skip
+				// blocks.
+				if ( $inner_content ) {
+					$null_indices  = array_keys( $inner_content, null, true );
+					$content_index = $null_indices[ $i ];
+					$nulls         = array_fill( 0, count( $blocks_to_insert ), null );
+					array_splice( $inner_content, $content_index, 1, $nulls );
+				}
+			} elseif ( ! empty( $blocks[ $i ]['innerBlocks'] ) ) {
+				$blocks[ $i ]['innerBlocks'] = gutenberg_replace_pattern_blocks(
```

Comment:
Is it possible to guard against recursions. Like a pattern block containing the same pattern block or something like that. cc @mcsf has some experience here.

Reply:
Right, I guess we can pass down the parent pattern name and check